File: analysis/constraints.py
# Ethan Doughty
# constraints.py
"""Constraint tracking and path-sensitive joins for dimension equality."""

from __future__ import annotations
from typing import Optional

from runtime.shapes import Dim, Shape
from runtime.symdim import SymDim
from runtime.env import Env


def _dim_key(d: Dim) -> tuple:
    """Structural sort key for canonicalizing constraint pairs.

    Args:
        d: Dimension (int, SymDim, or None)

    Returns:
        Sort key tuple: (type_tag, value) for stable ordering
        - int: (0, value)
        - SymDim: (1, terms_tuple)
        - None: (2,)
    """
    if isinstance(d, int):
        return (0, d)
    elif isinstance(d, SymDim):
        return (1, d._terms)
    elif d is None:
        return (2,)
    else:
        # Should not reach here
        return (3,)


def record_constraint(ctx, env: Env, dim1: Dim, dim2: Dim, line: int) -> None:
    """Record equality constraint between two dimensions.

    Skips constraints that are:
    - Involving None (unknown can't participate)
    - Trivial (dim1 == dim2)
    - Both concrete ints (already caught by dims_definitely_conflict)
    - Involving pre-bound variables (vars already bound in env)

    Args:
        ctx: AnalysisContext with constraints and constraint_provenance
        env: Current environment for checking pre-bound variables
        dim1: First dimension
        dim2: Second dimension
        line: Source line number (for provenance)
    """

    # Skip None dims
    if dim1 is None or dim2 is None:
        return

    # Skip trivial equality
    if dim1 == dim2:
        return

    # Skip both-concrete (already caught by dims_definitely_conflict)
    if isinstance(dim1, int) and isinstance(dim2, int):
        return

    # Check for pre-bound vars: if dim is simple var and var is bound in env, skip
    def is_prebound_var(d: Dim) -> bool:
        """Check if dimension is a simple variable that's already bound."""
        if not isinstance(d, SymDim):
            return False
        # Check if SymDim is a simple variable (one term with coeff=1)
        if len(d._terms) != 1:
            return False
        mono, coeff = d._terms[0]
        # Check coefficient is 1
        if coeff != 1:
            return False
        # Check monomial is single variable with exponent 1
        if len(mono) != 1:
            return False
        var_name, exp = mono[0]
        if exp != 1:
            return False
        # Check if var_name is bound in env
        return not env.get(var_name).is_bottom()

    if is_prebound_var(dim1) or is_prebound_var(dim2):
        return

    # Canonicalize: sort by _dim_key to ensure (a, b) and (b, a) are same
    key1 = _dim_key(dim1)
    key2 = _dim_key(dim2)
    if key1 <= key2:
        canonical = (dim1, dim2)
    else:
        canonical = (dim2, dim1)

    # Add to constraints set
    ctx.constraints.add(canonical)

    # Store provenance (first-seen: keep original line, not last writer)
    ctx.constraint_provenance.setdefault(canonical, line)


def snapshot_constraints(ctx) -> frozenset:
    """Return a copy of current constraint set.

    Args:
        ctx: AnalysisContext with constraints

    Returns:
        Frozen copy of current constraints
    """
    from analysis.context import AnalysisContext
    return frozenset(ctx.constraints)


def join_constraints(baseline: frozenset, branch_sets: list) -> set:
    """Path-sensitive join: keep baseline + constraints added in ALL branches.

    Args:
        baseline: Constraints before branching
        branch_sets: List of constraint sets from each branch (frozenset or set)

    Returns:
        Joined constraint set (baseline + common new constraints)
    """
    # Extract new constraints per branch
    new_per_branch = [set(branch) - set(baseline) for branch in branch_sets]

    # Intersection of all new constraints (only if all branches added them)
    if not new_per_branch:
        common_new = set()
    else:
        common_new = set(new_per_branch[0])
        for new_set in new_per_branch[1:]:
            common_new &= new_set

    # Return baseline + common new
    return set(baseline) | common_new


def try_extract_const_value(expr) -> Optional[int]:
    """Extract integer value from a Const expression if possible.

    Args:
        expr: Expression node from IR

    Returns:
        Integer value if expr is Const with integer value, else None
    """
    from ir import Const

    if not isinstance(expr, Const):
        return None

    # Check if value is an integer (or float with .0)
    value = expr.value
    if isinstance(value, (int, float)):
        if float(value).is_integer():
            return int(value)

    return None


def validate_binding(ctx, env, var_name: str, value: int, warnings: list, line: int) -> None:
    """Check if binding var_name=value conflicts with recorded constraints.

    Args:
        ctx: AnalysisContext with constraints and constraint_provenance
        env: Current environment (to look up other variables)
        var_name: Variable name being bound to concrete value
        value: Concrete integer value being assigned
        warnings: List to append warnings to
        line: Source line number of binding
    """
    from analysis.context import AnalysisContext
    import analysis.diagnostics as diag

    # Create target dimension for this variable
    target_dim = SymDim.var(var_name)

    # Search all constraints for ones involving this variable
    for dim1, dim2 in ctx.constraints:
        # Check if this constraint involves our target variable
        other_dim = None
        if dim1 == target_dim:
            other_dim = dim2
        elif dim2 == target_dim:
            other_dim = dim1

        if other_dim is None:
            continue

        # Found a constraint involving var_name
        # Check if other_dim conflicts with value
        conflict_value = None

        if isinstance(other_dim, int):
            # Other side is concrete int
            if other_dim != value:
                conflict_value = other_dim
        elif isinstance(other_dim, SymDim):
            # Other side is symbolic — check if it's a known constant
            const_val = other_dim.const_value()
            if const_val is not None and const_val != value:
                conflict_value = other_dim
            else:
                # other_dim is a symbolic variable — check if it's been bound to a scalar
                # Extract variable name from SymDim if it's a simple variable
                other_vars = other_dim.variables()
                if len(other_vars) == 1:
                    # Simple variable — check if bound in ctx.scalar_bindings
                    other_var_name = list(other_vars)[0]
                    if other_var_name in ctx.scalar_bindings:
                        other_value = ctx.scalar_bindings[other_var_name]
                        if other_value != value:
                            conflict_value = other_value

        if conflict_value is not None:
            # Emit conflict warning
            source_line = ctx.constraint_provenance.get((dim1, dim2), 0)
            warnings.append(diag.warn_constraint_conflict(
                line, var_name, value, conflict_value, source_line
            ))

File: src/Constraints.fs
module Constraints

open Ir
open Shapes
open Env
open SymDim
open DimEquiv
open Context

// ---------------------------------------------------------------------------
// Constraint tracking and path-sensitive joins for dimension equality.
// Port of analysis/constraints.py
// ---------------------------------------------------------------------------

/// tryExtractConstValue: extract integer value from a Const expression, or None.
/// Port of analysis/constraints.py try_extract_const_value().
let tryExtractConstValue (expr: Expr) : int option =
    match expr with
    | Const(_, v) when v = System.Math.Floor(v : float) -> Some (int v)
    | _ -> None


/// dimKey: structural sort key for canonicalizing constraint pairs.
let private dimKey (d: Dim) : int * string =
    match d with
    | Concrete n -> (0, string n)
    | Symbolic s -> (1, SymDim.toString s)
    | Range _    -> (3, Shapes.dimStr d)   // Range sorts after Symbolic
    | Unknown    -> (2, "")


/// isPreBoundVar: check if a Dim is a simple variable that's already bound in env.
let private isPreBoundVar (d: Dim) (env: Env) : bool =
    match d with
    | Symbolic s ->
        // Check if s is a simple variable: one term, coeff=1, single var with exp=1
        match s._terms with
        | [([varName, 1], coeff)] when coeff = Rational.One ->
            // Check if varName is bound in env (not bottom)
            not (Env.get env varName = Bottom)
        | _ -> false
    | Range _ -> false   // Range dims are not simple pre-bound variables
    | _ -> false


/// recordConstraint: record equality constraint between two dimensions.
/// Skips trivial, both-concrete, None, and pre-bound constraints.
let recordConstraint (ctx: AnalysisContext) (env: Env) (dim1: Dim) (dim2: Dim) (line: int) : unit =
    match dim1, dim2 with
    | Unknown, _ | _, Unknown -> ()
    | _ when dim1 = dim2 -> ()
    | Concrete _, Concrete _ -> ()   // already caught by dimsDefinitelyConflict
    | _ ->
        if isPreBoundVar dim1 env || isPreBoundVar dim2 env then ()
        else
            // Canonicalize: sort by dimKey to ensure (a, b) and (b, a) are same
            let k1 = dimKey dim1
            let k2 = dimKey dim2
            let canonKey1, canonKey2 =
                if k1 <= k2 then (Shapes.dimStr dim1, Shapes.dimStr dim2)
                else (Shapes.dimStr dim2, Shapes.dimStr dim1)
            let canonical = (canonKey1, canonKey2)

            ctx.cst.constraints.Add(canonical) |> ignore
            // Store provenance (first-seen: keep original line)
            if not (ctx.cst.constraintProvenance.ContainsKey(canonical)) then
                ctx.cst.constraintProvenance.[canonical] <- line

            // DimEquiv: union the two dimension strings
            let key1 = Shapes.dimStr dim1
            let key2 = Shapes.dimStr dim2
            DimEquiv.union ctx.cst.dimEquiv key1 key2 |> ignore

            // Propagate concrete values into the equivalence class
            match dim1 with
            | Concrete n -> DimEquiv.setConcrete ctx.cst.dimEquiv key1 n |> ignore
            | _ -> ()
            match dim2 with
            | Concrete n -> DimEquiv.setConcrete ctx.cst.dimEquiv key2 n |> ignore
            | _ -> ()


/// resolveDim: resolve a symbolic dim using the equivalence store.
/// 1. Check if the whole dim's key has a concrete value in the equiv store.
/// 2. If not, extract individual variables, substitute known concrete values.
/// 3. If all variables resolve, evaluate to Concrete; partial resolution returns simplified Symbolic.
let resolveDim (ctx: AnalysisContext) (d: Dim) : Dim =
    match d with
    | Symbolic s ->
        let key = Shapes.dimStr d
        match DimEquiv.getConcrete ctx.cst.dimEquiv key with
        | Some n -> Concrete n
        | None ->
            // Substitute individual variables from equiv classes
            let vars = SymDim.SymDim.variables s |> Set.toList
            let bindings =
                vars |> List.choose (fun v ->
                    match DimEquiv.getConcrete ctx.cst.dimEquiv v with
                    | Some n -> Some (v, n)
                    | None -> None)
                |> Map.ofList
            if Map.isEmpty bindings then d
            else
                match SymDim.SymDim.evaluate bindings s with
                | Some n -> Concrete n
                | None ->
                    let sub = SymDim.SymDim.substitute bindings s
                    match SymDim.SymDim.constValue sub with
                    | Some cv -> Concrete cv
                    | None -> Symbolic sub
    | _ -> d


/// resolveShape: apply resolveDim to all dims in a Shape.
let resolveShape (ctx: AnalysisContext) (shape: Shape) : Shape =
    match shape with
    | Matrix(r, c) ->
        let r' = resolveDim ctx r
        let c' = resolveDim ctx c
        if r' = r && c' = c then shape else Matrix(r', c')
    | Cell(r, c, elems) ->
        let r' = resolveDim ctx r
        let c' = resolveDim ctx c
        if r' = r && c' = c then shape else Cell(r', c', elems)
    | _ -> shape


/// snapshotConstraints: return a copy of current constraint set.
let snapshotConstraints (ctx: AnalysisContext) : System.Collections.Generic.HashSet<string * string> =
    System.Collections.Generic.HashSet<string * string>(ctx.cst.constraints)


/// joinConstraints: path-sensitive join: keep baseline + constraints added in ALL branches.
let joinConstraints
    (baseline: System.Collections.Generic.HashSet<string * string>)
    (branchSets: System.Collections.Generic.HashSet<string * string> list)
    : System.Collections.Generic.HashSet<string * string> =

    // Extract new constraints per branch (branch - baseline)
    let newPerBranch =
        branchSets
        |> List.map (fun branch ->
            let newSet = System.Collections.Generic.HashSet<string * string>(branch)
            for b in baseline do newSet.Remove(b) |> ignore
            newSet)

    // Intersection of all new constraints (only if all branches added them)
    let commonNew =
        match newPerBranch with
        | [] -> System.Collections.Generic.HashSet<string * string>()
        | first :: rest ->
            let result = System.Collections.Generic.HashSet<string * string>(first)
            for s in rest do
                result.IntersectWith(s)
            result

    // Return baseline + common new
    let result = System.Collections.Generic.HashSet<string * string>(baseline)
    for c in commonNew do result.Add(c) |> ignore
    result


/// validateBinding: check if binding var_name=value conflicts with recorded constraints.
let validateBinding
    (ctx: AnalysisContext)
    (env: Env)
    (varName: string)
    (value: int)
    (warnings: ResizeArray<Diagnostics.Diagnostic>)
    (line: int)
    : unit =

    let targetDimStr = Shapes.dimStr (Symbolic (SymDim.SymDim.var varName))

    for (d1, d2) in ctx.cst.constraints do
        // Check if this constraint involves our target variable
        let otherDimStr =
            if d1 = targetDimStr then Some d2
            elif d2 = targetDimStr then Some d1
            else None

        match otherDimStr with
        | None -> ()
        | Some other ->
            // Found a constraint: check if 'other' conflicts with value
            // Try parse 'other' as integer
            match System.Int32.TryParse(other) with
            | true, otherInt ->
                if otherInt <> value then
                    let sourceLine =
                        match ctx.cst.constraintProvenance.TryGetValue((d1, d2)) with
                        | true, l -> l
                        | _ -> 0
                    warnings.Add(Diagnostics.warnConstraintConflict line varName value other sourceLine)
            | _ ->
                // Check if other dim is in scalar_bindings
                match ctx.cst.scalarBindings.TryGetValue(other) with
                | true, otherValue when otherValue <> value ->
                    let sourceLine =
                        match ctx.cst.constraintProvenance.TryGetValue((d1, d2)) with
                        | true, l -> l
                        | _ -> 0
                    warnings.Add(Diagnostics.warnConstraintConflict line varName value other sourceLine)
                | _ -> ()
